File: projects/python/05-lru-cache/solution/main.py
# 05-lru-cache (solution)
import sys

class Node:
    def __init__(self, key: str = "", val: int = 0):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: Node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_head(self, node: Node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def get(self, key: str) -> int:
        if key not in self.map:
            return -1
        node = self.map[key]
        self._remove(node)
        self._add_to_head(node)
        return node.val

    def put(self, key: str, val: int):
        if key in self.map:
            node = self.map[key]
            node.val = val
            self._remove(node)
            self._add_to_head(node)
        else:
            if len(self.map) >= self.cap:
                lru = self.tail.prev
                self._remove(lru)
                del self.map[lru.key]
            node = Node(key, val)
            self.map[key] = node
            self._add_to_head(node)

def main():
    cache = None
    for line in sys.stdin:
        parts = line.strip().split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == "cap":
            cache = LRUCache(int(parts[1]))
        elif cmd == "put" and cache:
            cache.put(parts[1], int(parts[2]))
        elif cmd == "get" and cache:
            print(cache.get(parts[1]))

if __name__ == "__main__":
    main()

File: projects/rust/05-lru-cache/solution/main.rs
// 05-lru-cache (solution)
#![allow(unused_variables, dead_code, unused_mut, unused_imports)]
use std::collections::HashMap;
use std::io::{self, BufRead};

struct SimpleLRU {
    cap: usize,
    map: HashMap<String, i64>,
    order: Vec<String>,
}

impl SimpleLRU {
    fn new(cap: usize) -> Self {
        SimpleLRU {
            cap,
            map: HashMap::new(),
            order: Vec::new(),
        }
    }

    fn get(&mut self, key: &str) -> Option<i64> {
        if let Some(&val) = self.map.get(key) {
            if let Some(pos) = self.order.iter().position(|x| x == key) {
                self.order.remove(pos);
            }
            self.order.push(key.to_string());
            Some(val)
        } else {
            None
        }
    }

    fn put(&mut self, key: String, val: i64) {
        if self.map.contains_key(&key) {
            self.map.insert(key.clone(), val);
            if let Some(pos) = self.order.iter().position(|x| x == &key) {
                self.order.remove(pos);
            }
            self.order.push(key);
        } else {
            if self.map.len() >= self.cap {
                if !self.order.is_empty() {
                    let lru = self.order.remove(0);
                    self.map.remove(&lru);
                }
            }
            self.map.insert(key.clone(), val);
            self.order.push(key);
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut cache: Option<SimpleLRU> = None;

    for line in stdin.lock().lines() {
        if let Ok(l) = line {
            let trimmed = l.trim();
            if trimmed.is_empty() { continue; }
            let parts: Vec<&str> = trimmed.split_whitespace().collect();
            if parts[0] == "cap" && parts.len() >= 2 {
                let cap: usize = parts[1].parse().unwrap_or(2);
                cache = Some(SimpleLRU::new(cap));
            } else if parts[0] == "put" && parts.len() >= 3 {
                if let Some(c) = cache.as_mut() {
                    let val: i64 = parts[2].parse().unwrap_or(0);
                    c.put(parts[1].to_string(), val);
                }
            } else if parts[0] == "get" && parts.len() >= 2 {
                if let Some(c) = cache.as_mut() {
                    match c.get(parts[1]) {
                        Some(v) => println!("{}", v),
                        None => println!("-1"),
                    }
                }
            }
        }
    }
}
